lineax/_solver/tridiagonal.py

```diff
@@ -117,6 +99,83 @@ def allow_dependent_rows(self, operator):
         return False
 
 
+@partial(mlir.lower_fun, multiple_results=False)
+def _tridiagonal_solve_gpu_unbatched_lowering(
+    diagonal, lower_diagonal, upper_diagonal, vector, _
+):
+    return lax.linalg.tridiagonal_solve(
+        jnp.append(0.0, lower_diagonal),
+        diagonal,
+        jnp.append(upper_diagonal, 0.0),
+        vector[:, None],
+    ).flatten()
+
+
+@partial(mlir.lower_fun, multiple_results=False)
+def _tridiagonal_solve_cpu_lowering(
+    diagonal, lower_diagonal, upper_diagonal, vector, _
+):
+    return _tridiagonal_solve_lineax(
+        diagonal, lower_diagonal, upper_diagonal, vector, 1
+    )
+
+
+def _tridiagonal_solve_lineax(diagonal, lower_diagonal, upper_diagonal, vector, unroll):
+    #
+    # notation from: https://en.wikipedia.org/wiki/Tridiagonal_matrix_algorithm
+    # _p indicates prime, ie. `d_p` is the variable name for d' on wikipedia
+    #
+
+    size = len(diagonal)
+
+    def thomas_scan(prev_cd_carry, bd):
+        c_p, d_p, step = prev_cd_carry
+        # the index of `a` doesn't matter at step 0 as
+        # we won't use it at all. Same for `c` at final step
+        a_index = jnp.where(step > 0, step - 1, 0)
+        c_index = jnp.where(step < size, step, 0)
+
+        b, d = bd
+        a, c = lower_diagonal[a_index], upper_diagonal[c_index]
+        denom = b - a * c_p
+        new_d_p = (d - a * d_p) / denom
+        new_c_p = c / denom
+        return (new_c_p, new_d_p, step + 1), (new_c_p, new_d_p)
+
+    def backsub(prev_x_carry, cd_p):
+        x_prev, step = prev_x_carry
+        c_p, d_p = cd_p
+        x_new = d_p - c_p * x_prev
+        return (x_new, step + 1), x_new
+
+    # not a dummy init! 0 is the proper value for all of these
+    init_thomas = (0, 0, 0)
+    init_backsub = (0, 0)
+    diag_vec = (diagonal, vector)
+    _, cd_p = lax.scan(thomas_scan, init_thomas, diag_vec, unroll=unroll)
+    _, solution = lax.scan(backsub, init_backsub, cd_p, reverse=True, unroll=unroll)
+
+    return solution
+
+
+_tridiagonal_solve_p = core.Primitive("_tridiagonal_solve_lineax")
+_tridiagonal_solve_p.def_impl(_tridiagonal_solve_lineax)
+_tridiagonal_solve_p.def_abstract_eval(Tridiagonal.abstract_eval)
+
+mlir.register_lowering(
+    _tridiagonal_solve_p, _tridiagonal_solve_cpu_lowering, platform="cpu"
+)
+mlir.register_lowering(
+    _tridiagonal_solve_p, _tridiagonal_solve_gpu_unbatched_lowering, platform="cuda"
+)
+mlir.register_lowering(
+    _tridiagonal_solve_p, _tridiagonal_solve_gpu_unbatched_lowering, platform="rocm"
+)
+mlir.register_lowering(
+    _tridiagonal_solve_p,
+    mlir.lower_fun(_tridiagonal_solve_lineax, multiple_results=False),
```

Comment:
Style nit, `mlir.lower_fun` is used as a decorator on the other branches and as an inline call here.